rustls/src/verify.rs

```diff
@@ -32,50 +32,87 @@ static SUPPORTED_SIG_ALGS: SignatureAlgorithms = &[
     &webpki::RSA_PKCS1_3072_8192_SHA384,
 ];
 
-/// Marker types.  These are used to bind the fact some verification
-/// (certificate chain or handshake signature) has taken place into
-/// protocol states.  We use this to have the compiler check that there
-/// are no 'goto fail'-style elisions of important checks before we
-/// reach the traffic stage.
+// Marker types.  These are used to bind the fact some verification
+// (certificate chain or handshake signature) has taken place into
+// protocol states.  We use this to have the compiler check that there
+// are no 'goto fail'-style elisions of important checks before we
+// reach the traffic stage.
+//
+// These types are public, but cannot be directly constructed.  This
+// means their origins can be precisely determined by looking
+// for their `assertion` constructors.
+
+/// Zero-sized marker type representing verification of a signature.
 ///
-/// These types are public, but cannot be directly constructed.  This
-/// means their origins can be precisely determined by looking
-/// for their `assertion` constructors.
-pub struct HandshakeSignatureValid(());
+/// This struct must be constructed using [`assertion`](Self::assertion).
+/// ```compile_fail
+/// # use rustls::client::HandshakeSignatureValid;
+/// // Doesn't compile
+/// let _ = HandshakeSignatureValid;
+/// // Compiles
+/// let _ = HandshakeSignatureValid::assertion();
+/// ```
+#[non_exhaustive]
+#[derive(Debug)]
+pub struct HandshakeSignatureValid;
```

Comment:
Normally `non_exhaustive` isn't used for a structure that has no public fields. Instead `HandshakeSignatureValid(())` would be more idiomatic, right? Or am I behind the times?

Reply:
Yeah, that was also my hesitation. I think maybe this has the same effect, but I think it might be less idiomatic. So I  think we don't want to make changes that replace a single private field with a fieldless `#[non_exhaustive]` type.

(I'm guessing this PR may have been triggered by the unidiomatic way the assertion types instantiate their private fields with `0: ()`, which I think triggers some kind of warning? But this could just be fixed by initializing as `HandshakeSignatureValid(())`.)

Reply:
Yea fair enough. Now that I think about it `non_exhaustive` isn't as clear. I saw it in a book and thought it looked cleaner. Can I just remove all the changes other than deriving `Debug`?

Reply:
Sure, that should be fine.